### PR TITLE
include Linkdependencies

### PR DESCRIPTION
**Pull request by AnjaliAbraham**
Added changes to the application properties file under application conf to include a new property.
Changed the impactUtilities.groovy to include the linkdependencies

### CODE REVIEW

File: utilities/ImpactUtilities.groovy [hunk 1]
```diff
@@ -867,4 +878,48 @@ def sortFileList(list) {
...
+                    def linkDepName = logicalDep.getLname()
+                    def linkDepLogicalFile = metadataStore.getCollection(props.applicationCollectionName).getLogicalFiles(linkDepName)
+
+                    // Get the logical path for all the link dependencies returned
+                    linkDepLogicalFile.each { filePath ->
+                        // Link Dependency Files to be added
+                        def linkDepFile = filePath.getFile()
+
+                        if (linkDepFile != logicalFile) {
+                            buildLinkSet.add(linkDepFile)
```
Comment (dennis-behm) [inline utilities/ImpactUtilities.groovy:915]: We should only add files, that have a ScriptMapping ( so it is a proper sub-module)
```
if (ScriptMappings.getScriptName(linkDepFile)) ... 
```

File: utilities/ImpactUtilities.groovy [hunk 2]
```diff
@@ -867,4 +878,48 @@ def sortFileList(list) {
...
+                            if (props.verbose) println "** Adding link dep to buildLinkset $buildLinkSet"
```
Comment (dennis-behm) [inline utilities/ImpactUtilities.groovy:916]:
```suggestion
                            if (props.verbose) println "** Adding submodule $linkDepFile to build list as link dependency for $logicalFile"
```

File: utilities/ImpactUtilities.groovy [hunk 3]
```diff
@@ -169,6 +169,17 @@ def createImpactBuildList() {
 		}
 
 	}
+	//include statically called sub programs when the main program changes
+     if (props.addSubmodulesToBuildList ==  'true') {
+        Set<String> buildLinkSet = new HashSet<String>()
+
+        // Call addLinkDependencies to append link dependencies to buildSet
+        if (props.verbose) println "**Calculation of statically called sub module programs"
```
Comment (dennis-behm) [inline utilities/ImpactUtilities.groovy:177]:
```suggestion
        if (props.verbose) println "*** Perform analysis to add statically called sub modules to build list."
```

**Review by dennis-behm — CHANGES_REQUESTED** [review]
@AnjaliAbraham thank you very much for this new contribution - would you please have a look to my comments/suggestions?

File: utilities/ImpactUtilities.groovy [hunk 4]
```diff
@@ -169,6 +169,17 @@ def createImpactBuildList() {
 		}
 
 	}
+	//include statically called sub programs when the main program changes
+     if (props.addSubmodulesToBuildList ==  'true') {
```
[the remark this answers is not in the document]
Comment (AnjaliAbraham) [reply utilities/ImpactUtilities.groovy:173]: Suggested change accepted

File: utilities/ImpactUtilities.groovy [hunk 5]
```diff
@@ -169,6 +169,17 @@ def createImpactBuildList() {
...
+        Set<String> buildLinkSet = new HashSet<String>()
+
+        // Call addLinkDependencies to append link dependencies to buildSet
+        if (props.verbose) println "**Calculation of statically called sub module programs"
+        buildLinkSet = addLinkDependencies(buildSet, buildLinkSet)
+        if (props.verbose) println "** Buildlinkset file $buildLinkSet"
```
[the remark this answers is not in the document]
Comment (AnjaliAbraham) [reply utilities/ImpactUtilities.groovy:179]: Change accepted

File: utilities/ImpactUtilities.groovy [hunk 6]
```diff
@@ -867,4 +878,48 @@ def sortFileList(list) {
...
+
+/*
+ *  addLinkDependencies -
+ *  method to identify all statically called sub module programs when the main program changes
+ *
+ *  @return list of statically called sub modules
+ *
+ */
+def addLinkDependencies(buildSet, buildLinkSet) {
+    String OutputsCollectionName = props.applicationOutputsCollectionName
```
[the remark this answers is not in the document]
Comment (AnjaliAbraham) [reply utilities/ImpactUtilities.groovy:892]: suggestion accepted

File: utilities/ImpactUtilities.groovy (hunk 2 above)
Comment (AnjaliAbraham) [reply utilities/ImpactUtilities.groovy:916]: Suggestion accepted

File: utilities/ImpactUtilities.groovy (hunk 3 above)
Comment (AnjaliAbraham) [reply utilities/ImpactUtilities.groovy:177]: Suggestion accepted

File: utilities/ImpactUtilities.groovy [hunk 7]
```diff
@@ -169,6 +169,17 @@ def createImpactBuildList() {
...
+	//include statically called sub programs when the main program changes
+     if (props.addSubmodulesToBuildList ==  'true') {
+        Set<String> buildLinkSet = new HashSet<String>()
+
+        // Call addLinkDependencies to append link dependencies to buildSet
+        if (props.verbose) println "**Calculation of statically called sub module programs"
+        buildLinkSet = addLinkDependencies(buildSet, buildLinkSet)
+        if (props.verbose) println "** Buildlinkset file $buildLinkSet" 
+        buildSet.addAll(buildLinkSet)
+        if (props.verbose) println "** Combined buildset $buildSet"
```
[the remark this answers is not in the document]
Comment (AnjaliAbraham) [reply utilities/ImpactUtilities.groovy]: Change accepted

File: utilities/ImpactUtilities.groovy [hunk 8]
```diff
@@ -867,4 +878,48 @@ def sortFileList(list) {
...
+
+        if (logicalFiles) {
+            // Check if any logical files are found
+            // List all link dependencies for every main program that changes from the output collection. 
+            // This will return only the program name and not the absolute path
+            logicalFiles.each { logicalFileRecord ->
+                def dependencies = logicalFileRecord.getLogicalDependencies()
+
+                dependencies.each { logicalDep ->
+                    def linkDepName = logicalDep.getLname()
```
[the remark this answers is not in the document]
Comment (AnjaliAbraham) [reply utilities/ImpactUtilities.groovy:906]: check included

File: utilities/ImpactUtilities.groovy (hunk 1 above)
Comment (AnjaliAbraham) [reply utilities/ImpactUtilities.groovy]: check included

File: samples/MortgageApplication/application-conf/application.properties [hunk 9]
```diff
@@ -44,6 +44,11 @@ excludeFileList=.*,**/.*,**/*.properties,**/*.xml,**/*.groovy,**/*.json,**/*.md,
 # sample: skipImpactCalculationList=**/epsmtout.cpy,**/centralCopybooks/*.cpy
 skipImpactCalculationList=
 
+# flag to enable the zAppBuild capability to include sub modules with static calls, when a main program changes
```
Comment (dennis-behm) [inline samples/MortgageApplication/application-conf/application.properties:47]:
```suggestion
# Flag to expand impact analysis to automatically include all mandatory submodules to build list in pipeline builds with a DBB Metadatastore connection based on the intermediate build list after impact analysis
# This functionality queries and relies on the DBB Output collection of the application
# 
# The feature is specifically useful for applications using static calls to increase the isolation of the build environment
```

File: samples/MortgageApplication/application-conf/application.properties [hunk 10]
```diff
@@ -44,6 +44,11 @@ excludeFileList=.*,**/.*,**/*.properties,**/*.xml,**/*.groovy,**/*.json,**/*.md,
...
+# If the property is set to true, if a main program undergoes a change, all statically called sub programs are recompiled
+# Default: false
```
Comment (dennis-behm) [inline samples/MortgageApplication/application-conf/application.properties:49]:
```suggestion
# Default: false
# This property is setup as a DBB file property. This allows to enable the feature just for a subset of your build files
# Sample to inspect the intermediate build list add submodules for Cobol build files
# addSubmodulesToBuildList=true :: **/*.cbl
```

File: utilities/ImpactUtilities.groovy [hunk 11]
```diff
@@ -169,6 +169,16 @@ def createImpactBuildList() {
 		}
 
 	}
+	//include statically called sub programs when the main program changes
+     if (props.addSubmodulesToBuildList && props.addSubmodulesToBuildList.toBoolean()) {
```
Comment (dennis-behm) [inline utilities/ImpactUtilities.groovy:173]: Either you need to introduce another property to enable the feature, or you need to loop through the intermediate build list to understand if the additional analysis is required via the property `addSubmodulesToBuildList` for a set of files.

**Review by dennis-behm — CHANGES_REQUESTED** [review]
@AnjaliAbraham 

I noticed that have configured the new properties for the MortgageApplication only.
Please add it to the sample application-conf in  `samples/application-conf`. Please also update the README.md in this directory to mention the new properties and their purpose about the new feature in there.

I had added a couple of suggestions and comments.

File: utilities/ImpactUtilities.groovy (hunk 11 above)
Comment (AnjaliAbraham) [reply utilities/ImpactUtilities.groovy]: Dennis - I have made the changes to include perform the analysis for a set of files as well introduced logic to reduce the number of metadatastore hits. I have also moved the logic before the 'changed configuration' check.

File: utilities/ImpactUtilities.groovy [hunk 12]
```diff
@@ -122,7 +122,23 @@ def createImpactBuildList() {
...
+	    
+	    Set<String> buildLinkSet = new HashSet<String>() 
+	    buildSet.each { buildFile ->
+	         String addSubmodulesToBuildList = props.getFileProperty('addSubmodulesToBuildList', buildFile)
+	 
+	         //include statically called sub programs when the main program changes
+		
+		    if (addSubmodulesToBuildList != null && addSubmodulesToBuildList.toBoolean()) {
+			   // Call addLinkDependencies to append link dependencies to buildSet
+			   if (props.verbose) println "** Perform analysis to add statically called sub modules to build list."
```
Comment (dennis-behm) [inline utilities/ImpactUtilities.groovy:134]:
```suggestion
			   if (props.verbose) println "** Perform analysis to add statically called sub modules to build list for ${buildFile}."
```

File: samples/MortgageApplication/application-conf/application.properties [hunk 13]
```diff
@@ -44,6 +44,17 @@ excludeFileList=.*,**/.*,**/*.properties,**/*.xml,**/*.groovy,**/*.json,**/*.md,
...
+# Flag to expand impact analysis to automatically include all mandatory submodules to build list in pipeline builds with a DBB Metadatastore connection based on the intermediate build list after impact analysis
+# This functionality queries and relies on the DBB Output collection of the application
+# 
+# The feature is specifically useful for applications using static calls to increase the isolation of the build environment
+# If the property is set to true, if a main program undergoes a change, all statically called sub programs are recompiled
+# Default: false
+# This property is setup as a DBB file property. This allows to enable the feature just for a subset of your build files
+# Sample to inspect the intermediate build list add submodules for Cobol build files
+# addSubmodulesToBuildList=true :: **/*.cbl
+addSubmodulesToBuildList=false :: **/*.cbl
```
Comment (dennis-behm) [inline samples/MortgageApplication/application-conf/application.properties:56]: This is a nice feature that can be show-cased by the MortgageApplication. Can we enable it for the MortgageApplication?

```suggestion
addSubmodulesToBuildList=true :: **/*.cbl
```

**Review by dennis-behm — COMMENTED** [review]
Hi @AnjaliAbraham, thanks for this enhancement. Two minor comments - I think it would be nice to print the buildFile to the output in verbose mode and by default enable it for MortgageApplication

**Comment by AnjaliAbraham** [comment]
Hi Dennis - Print the buildFile to the output in verbose mode and enable it for MortgageApplication by default has been committed. Thanks

**Review by dennis-behm — APPROVED** [review]
Thank you @AnjaliAbraham 👍

**Pull request merged**